add pragma - node.py, core.py, scheduler.py

scharlottej13 opened the pull request:
- [ ] partially addresses #5445

GPUtester commented:
Can one of the admins verify this patch?

phobson reviewed (commented):
Overall I'm comfortable with 95% of this. I'd be interested to get @fjetter 's feedback on the changes to the .coveragerc file. 

My poorly calibrated instincts want to see tests on those files, mostly because the tests _seem_ fairly quick to write. I'll defer to more experienced judgement on that.

That question aside, this is a  👍🏻  for me.

scharlottej13 commented:
> Overall I'm comfortable with 95% of this. I'd be interested to get @fjetter 's feedback on the changes to the .coveragerc file.
> 
> My poorly calibrated instincts want to see tests on those files, mostly because the tests _seem_ fairly quick to write. I'll defer to more experienced judgement on that.
> 
> That question aside, this is a 👍🏻 for me.

Thanks @phobson! Florian requested the `.coveragerc` changes, I think since they already have 0% coverage. I'm still building intuition on this, though, so I may be mistaken.

fjetter commented on distributed/core.py:
```diff
@@ -474,7 +474,7 @@ async def handle_comm(self, comm):
                 try:
                     op = msg.pop("op")
-                except KeyError as e:
+                except KeyError as e:  # pragma: no cover
```
I have no idea what would actually happen if this exception is raised, i.e. what will the server do? Just continue working? Block? Close?
I think this should be tested

```suggestion
                except KeyError as e:
```

fjetter commented on distributed/node.py:
```diff
@@ -164,7 +164,7 @@ def start_http_server(
                     change_port = True
                     retries_left = retries_left - 1
-                    if retries_left < 1:
+                    if retries_left < 1:  # pragma: no cover
```
This is pretty important piece for the retry logic. Will it retry once? Will it retry forever? Will it retry for Xs?
It is rather trivial to understand what these two lines are doing but this behaviour can easily be lost in a refactor and a test should protect us. It's fine to change this behaviour but it should be done deliberately

```suggestion
                    if retries_left < 1:
```

fjetter commented on distributed/scheduler.py:
```diff
@@ -2753,7 +2753,7 @@ def transition_waiting_processing(self, key):
 
             return recommendations, client_msgs, worker_msgs
-        except Exception as e:
+        except Exception as e:  # pragma: no cover
```
Just curious but can the omit_lines thing be a multiline regex?

`except\sException\sas\s(\w*):\n*\s*logger.exception\(\1\)\n*\s*if\sLOG_PDB:`

should match this and we could remove all of the extra pragmas

scharlottej13 replied:
Just tested this locally and doesn't look like it :/. Also found some SO posts on the topic, but seems the exceptions are designed to be executed line by line.

fjetter replied:
thanks for checking

fjetter commented on distributed/scheduler.py:
```diff
@@ -4474,7 +4474,7 @@ async def add_worker(
                     if inspect.isawaitable(result):
                         await result
-                except Exception as e:
+                except Exception as e:  # pragma: no cover
```
I do think we should test failing plugins even though its not top priority it is important for users to know what happens if their plugin errs, e.g. does the cluster come up? Will the exception only be logged? Something undefined?

```suggestion
                except Exception as e:
```

fjetter commented on distributed/scheduler.py:
```diff
@@ -4896,7 +4896,7 @@ def update_graph(
                     annotations=annotations,
                 )
-            except Exception as e:
+            except Exception as e:  # pragma: no cover
```
see above
```suggestion
            except Exception as e:
```

fjetter commented on distributed/scheduler.py:
```diff
@@ -5097,7 +5097,7 @@ async def remove_worker(self, comm=None, address=None, safe=False, close=True):
                     if inspect.isawaitable(result):
                         await result
-                except Exception as e:
+                except Exception as e:  # pragma: no cover
```
see above
```suggestion
                except Exception as e:
```

fjetter commented on distributed/scheduler.py:
```diff
@@ -5388,7 +5388,7 @@ def report(self, msg: dict, ts: TaskState = None, client: str = None):
                 c.send(msg)
                 # logger.debug("Scheduler sends message to client %s", msg)
-            except CommClosedError:
+            except CommClosedError:  # pragma: no cover
```
This is connected to failing batched sends which is smoething we do not have a proper solution for. There are a bunch of issues open and having this tested is important (even though we don't have a nice solution for it, yet)

```suggestion
            except CommClosedError:
```

fjetter commented on distributed/scheduler.py:
```diff
@@ -5409,7 +5409,7 @@ async def add_client(self, comm, client=None, versions=None):
             try:
                 plugin.add_client(scheduler=self, client=client)
-            except Exception as e:
+            except Exception as e:  # pragma: no cover
```
see above
```suggestion
            except Exception as e:
```

fjetter commented on distributed/scheduler.py:
```diff
@@ -5464,7 +5464,7 @@ def remove_client(self, client=None):
                 try:
                     plugin.remove_client(scheduler=self, client=client)
-                except Exception as e:
+                except Exception as e:  # pragma: no cover
```
see above
```suggestion
                except Exception as e:
```

fjetter commented on distributed/scheduler.py:
```diff
@@ -5483,15 +5483,15 @@ def send_task_to_worker(self, worker, ts: TaskState, duration: double = -1):
             raise
 
-    def handle_uncaught_error(self, **msg):
+    def handle_uncaught_error(self, **msg):  # pragma: no cover
         logger.exception(clean_exception(**msg)[1])
```
This is dead code. Please remove the code

fjetter commented on distributed/scheduler.py:
```diff
@@ -5575,7 +5575,7 @@ def handle_long_running(self, key=None, worker=None, compute_duration=None):
         """
         parent: SchedulerState = cast(SchedulerState, self)
-        if key not in parent._tasks:
+        if key not in parent._tasks:  # pragma: no cover
             logger.debug("Skipping long_running since key %s was already released", key)
```
I'm not sure if the current behaviour is stable
```suggestion
        if key not in parent._tasks:
```

fjetter commented on distributed/scheduler.py:
```diff
@@ -5805,7 +5805,7 @@ def worker_send(self, worker, msg):
         try:
             stream_comms[worker].send(msg)
-        except (CommClosedError, AttributeError):
+        except (CommClosedError, AttributeError):  # pragma: no cover
```
```suggestion
        except (CommClosedError, AttributeError):            
```

fjetter commented on distributed/scheduler.py:
```diff
@@ -7130,7 +7130,7 @@ async def feed(
         eventually be phased out.  It is mostly used by diagnostics.
         """
-        if not dask.config.get("distributed.scheduler.pickle"):
+        if not dask.config.get("distributed.scheduler.pickle"):  # pragma: no cover
```
This is an actual feature. We should test it

```suggestion
        if not dask.config.get("distributed.scheduler.pickle"):
```

scharlottej13 commented:
Just addressed your comments, thanks for the review @phobson @fjetter and for taking the time to write out your explanations in removing pragma statements!
